indico/modules/rb/client/js/modules/admin/MapAreasPage.jsx

```diff
@@ -0,0 +1,355 @@
+        areas: PropTypes.arrayOf(PropTypes.object).isRequired,
+        actions: PropTypes.exact({
+            createMapArea: PropTypes.func.isRequired,
+            updateMapAreas: PropTypes.func.isRequired,
+            deleteMapAreas: PropTypes.func.isRequired,
+            fetchMapAreas: PropTypes.func.isRequired,
+        }).isRequired,
+    };
+
+    constructor(props) {
+        super(props);
+
+        this.map = null;
+        this.featureGroup = null;
+    }
+
+    state = {
+        areaProps: null,
+        suppressClick: false
+    };
+
+    componentDidMount() {
+        const {actions: {fetchMapAreas}} = this.props;
+        fetchMapAreas();
+        // rerender is required here because only at this point this.map
+        // refers to the actual map element which is then necessary in this.drawAreas
+        this.forceUpdate();
+    }
+
+    setMapAreaRef = ref => {
+        this.map = ref;
+    };
+
+    setFeatureGroupRef = ref => {
+        this.featureGroup = ref;
+    };
+
+    createNewArea = (evt) => {
+        const {
+            layerType,
+            layer: {
+                _bounds: {_northEast: northEast, _southWest: southWest}
+            }
+        } = evt;
+
+        if (layerType !== 'rectangle') {
+            return;
+        }
+
+        this.setState({
+            areaProps: {bounds: {northEast, southWest}}
+        });
+
+        if (this.featureGroup) {
+            // this is necessary here so that we don't end up with
+            // the initial layer being still visible on the map
+            this.featureGroup.leafletElement.eachLayer((layer) => {
+                const {options: {id}} = layer;
+                if (id === undefined) {
+                    this.featureGroup.leafletElement.removeLayer(layer);
+                }
+            });
+        }
+    };
+
+    editArea = (evt) => {
+        if (!Object.keys(evt.layers._layers).length) {
+            return;
+        }
+
+        const {actions: {updateMapAreas}} = this.props;
+        const editedAreas = [];
+
+        evt.layers.eachLayer(({options: {id}, _bounds}) => {
+            const {_southWest: southWest, _northEast: northEast} = _bounds;
+            editedAreas.push({
+                id,
+                bounds: {southWest, northEast}
+            });
+        });
+
+        updateMapAreas(editedAreas);
+    };
+
+    deleteMapAreas = (evt) => {
+        if (!Object.keys(evt.layers._layers).length) {
+            return;
+        }
+
+        const {actions: {deleteMapAreas}} = this.props;
+        const deletedAreas = [];
+
+        evt.layers.eachLayer(({options: {id}}) => {
+            deletedAreas.push(id);
+        });
+
+        deleteMapAreas(deletedAreas);
+    };
+
+    drawAreas = () => {
+        if (this.map === null) {
+            return null;
+        }
+
+        const {areas} = this.props;
+        return areas.map((area) => {
+            const southWest = [area.top_left_latitude, area.top_left_longitude];
```

Comment:
GIven the amount of times you write `area`, I would consider de-structuring in the function args already.
Actually, this could be a good candidate for a `_drawArea` method. It's a bit of a massive method right now.